open-craft/xblock-html-completable
# [BB-1780] Create completable HTML XBlock

This creates a new XBlock that disables automatic `publish_completion` event and enables custom `complete` endpoint that uses `emit_completion` from `CompletableXBlockMixin`.

Example use-case: we have a Wistia block and we want to add a custom completion even when the user has the certain amount of the video watched. To do this, we can embed the following script in the completable HTML XBlock:
   ```html
   <script>
   // Wait for the page to load.
   window.addEventListener('load', function () {
     window._wq = window._wq || [];
       _wq.push({ "_all": function(video) {
         video.bind('percentwatchedchanged', function (percent, lastPercent) {
           if (percent >= .25 && lastPercent < .25) {
             console.log('The viewer has watched 25% of the video!')
             document.getElementsByClassName("xblock-student_view-completable_html5")[0].submitCompletion(); // Trigger `emit_completion`.
           }
         });
     }});
   });
   </script>
   ```
# Testing instructions:
1. Go to http://localhost:18000/admin/waffle/switch/ and set `completion.enable_completion_tracking` as `active`.
1. Install this version of `xblock-html` (`make lms/studio-shell`, `pip install -e /edx/src/xblock-html-completable/`, `make lms-restart && make studio-restart`).
1. Enable `completable_html5` in advanced settings of a course.
1. Create `Advanced -> Completable` component.
1. In the Editor paste the following:
   ```html
      <p><strong>This is just a simple test.</strong></p>
      <script>
         // Wait for the page to fully load.
         window.addEventListener('load', function () {
            document.getElementsByClassName("xblock-student_view-completable_html5")[0].submitCompletion();
         });
      </script>
   ```
1. Publish the block and visit it int the LMS.
1. Go to the Django shell (`make lms-shell`, `/edx/bin/edxapp-shell-lms`) and type:
   ```python
     from completion.models import BlockCompletion
     BlockCompletion.objects.all()
     BlockCompletion.objects.all().delete()
   ```
1. Check that the `complete` POST is sent and completion appears in Django shell (and in the UI after refreshing the page).
1. (Optional) Remove the `click` from the HTML block's content and see that the `complete` request is no longer POSTed.
